src/main/java/com/google/sps/servlets/GoDataServlet.java

```diff
@@ -14,108 +14,89 @@
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.KeyFactory;
 import com.google.gson.Gson;
-import com.google.sps.data.HuntItem;
-import com.google.sps.data.LatLng;
-import com.google.sps.data.Riddle;
 import com.google.sps.data.ScavengerHunt;
 import java.io.IOException;
-import java.util.ArrayList;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
-/** Returns a fake scavenger hunt and updates the current index of the hunt. */
+/** Returns a scavenger hunt and updates the current index of the hunt. */
 @WebServlet("/go-data")
 public class GoDataServlet extends HttpServlet {
   private static final String INDEX_PARAMETER = "new-index";
 
-  // Keeps track of the current stage in the hunt that the user is on.
-  // Before the user begins the hunt, the index should be -1.
-  private int index = -1;
+  private static final DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
 
   /**
    * Updates the index of the scavenger hunt (aka the destination that the user currently needs to
-   * find.
+   * find).
    */
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    Gson gson = new Gson();
     String indexStr = request.getParameter(INDEX_PARAMETER);
+    String huntIDStr = request.getParameter(Constants.HUNTID_PARAMETER);
     try {
-      index = Integer.parseInt(indexStr);
+      int index = Integer.parseInt(indexStr);
+      long huntID = Long.parseLong(huntIDStr);
+      Entity huntEntity = findScavengerHunt(huntID);
+      if (huntEntity == null) {
+        return;
+      }
+
+      // Update index of scavenger hunt.
+      ScavengerHunt hunt =
+          gson.fromJson((String) huntEntity.getProperty(Constants.HUNT_VAL), ScavengerHunt.class);
+      hunt.updateIndex(index);
+
+      // Convert hunt to JSON string and put into Datastore
+      String huntStr = gson.toJson(hunt);
+      huntEntity.setProperty(Constants.HUNT_VAL, huntStr);
+      datastore.put(huntEntity);
     } catch (Exception e) {
     }
 
     // Redirect back to main page.
     response.sendRedirect(Constants.GO_URL);
   }
 
+  /** Retrieves scavenger hunt data from Datastore, and sends to /go-data. */
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    ScavengerHunt hunt = buildScavengerHunt();
+    String huntIDStr = request.getParameter(Constants.HUNTID_PARAMETER);
+    Entity huntEntity = null;
+    try {
+      long huntID = Long.parseLong(huntIDStr);
+      huntEntity = findScavengerHunt(huntID);
+    } catch (Exception e) {
+    }
+
+    if (huntEntity == null) {
+      return;
```

Comment:
what happens on the client side if the huntEntity is null?  The user might want to see an error message or something, right?

Reply:
I just set the response to an error message within this if loop.